optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java

```diff
@@ -712,6 +714,23 @@ public ScoreDefinition getScoreDefinition() {
         return memberNames;
     }
 
+    public Set<Class<?>> getProblemFactClassSet() {
+        Stream<Class<?>> factClassStream = getProblemFactMemberAccessorMap()
+                .values()
+                .stream()
+                .map(MemberAccessor::getType);
+        Stream<Class<?>> factCollectionClassStream = getProblemFactCollectionMemberAccessorMap()
+                .entrySet()
+                .stream()
+                .map(entry -> {
+                    MemberAccessor accessor = entry.getValue();
+                    return ConfigUtils.extractCollectionGenericTypeParameter("solutionClass", getSolutionClass(),
+                            accessor.getType(), accessor.getGenericType(), ProblemFactCollectionProperty.class, entry.getKey());
```

Comment:
I think the problemFactClassSet should be a cached as a field on the SolutionDescriptor, so this code runs once.
This code to fill in that new field could run at the end of `SolutionDescriptor.processProblemFactPropertyAnnotation()` (or at the end of `SolutionDescriptor.processFactEntityOrScoreAnnotation()` if it relies on entities too).

A basic rule around reflection is to scan each class only once for the same thing.
That rule is violated as this code calls `extractCollectionGenericTypeParameter()` (which uses reflection) for every from() call.
Also, as this code calls that reflection method, thinking Quarkus wise, it's much better if it already runs during SolutinDescriptor creation.